numba/_helperlib.c

```diff
@@ -17,6 +17,301 @@
     #define NPY_ARRAY_BEHAVED NPY_BEHAVED
 #endif
 
+/*
+ * PRNG support.
+ */
+
+/* Magic Mersenne Twister constants */
+#define MT_N 624
+#define MT_M 397
+#define MT_MATRIX_A 0x9908b0dfU
+#define MT_UPPER_MASK 0x80000000U
+#define MT_LOWER_MASK 0x7fffffffU
+
+/* unsigned int is sufficient on modern machines as we only need 32 bits */
+typedef struct {
+    int index;
+    unsigned int mt[MT_N];
+    int has_gauss;
+    double gauss;
+} rnd_state_t;
+
+static rnd_state_t py_random_state;
+static rnd_state_t np_random_state;
+
+/* Some code portions below from CPython's _randommodule.c, some others
+   from Numpy's and Jean-Sebastien Roy's randomkit.c. */
+
+static void
+Numba_rnd_shuffle(rnd_state_t *state)
+{
+    int i;
+    unsigned int y;
+
+    for (i = 0; i < MT_N - MT_M; i++) {
+        y = (state->mt[i] & MT_UPPER_MASK) | (state->mt[i+1] & MT_LOWER_MASK);
+        state->mt[i] = state->mt[i+MT_M] ^ (y >> 1) ^ (-(y & 1) & MT_MATRIX_A);
+    }
+    for (; i < MT_N - 1; i++) {
+        y = (state->mt[i] & MT_UPPER_MASK) | (state->mt[i+1] & MT_LOWER_MASK);
+        state->mt[i] = state->mt[i+(MT_M-MT_N)] ^ (y >> 1) ^ (-(y & 1) & MT_MATRIX_A);
+    }
+    y = (state->mt[MT_N - 1] & MT_UPPER_MASK) | (state->mt[0] & MT_LOWER_MASK);
+    state->mt[MT_N - 1] = state->mt[MT_M - 1] ^ (y >> 1) ^ (-(y & 1) & MT_MATRIX_A);
+}
+
+/* Initialize mt[] with an integer seed */
+static void
+Numba_rnd_init(rnd_state_t *state, unsigned int seed)
+{
+    unsigned int pos;
+    seed &= 0xffffffffU;
+
+    /* Knuth's PRNG as used in the Mersenne Twister reference implementation */
+    for (pos = 0; pos < MT_N; pos++) {
+        state->mt[pos] = seed;
+        seed = (1812433253U * (seed ^ (seed >> 30)) + pos + 1) & 0xffffffffU;
+    }
+    state->index = MT_N;
+    state->has_gauss = 0;
+}
+
+/* Random-initialize the given state (for use at startup) */
+static int
+_rnd_random_seed(rnd_state_t *state)
+{
+    PyObject *timemod, *timeobj;
+    timemod = PyImport_ImportModuleNoBlock("time");
+    double time;
+    unsigned int seed, rshift;
+
+    if (!timemod)
+        return -1;
+    timeobj = PyObject_CallMethod(timemod, "time", NULL);
+    Py_DECREF(timemod);
+    time = PyFloat_AsDouble(timeobj);
+    Py_DECREF(timeobj);
+    if (time == -1 && PyErr_Occurred())
+        return -1;
+    /* Mix in seconds and nanoseconds */
+    seed = (unsigned) time ^ (unsigned) (time * 1e9);
+#ifndef _WIN32
+    seed ^= getpid();
+#endif
+    /* Address space randomization bits: take MSBs of various pointers.
+     * It is counter-productive to shift by 32, since the virtual address
+     * space width is usually less than 64 bits (48 on x86-64).
+     */
+    rshift = sizeof(void *) > 4 ? 16 : 0;
+    seed ^= (Py_uintptr_t) &timemod >> rshift;
+    seed += (Py_uintptr_t) &PyObject_CallMethod >> rshift;
+    Numba_rnd_init(state, seed);
```

Comment:
Ultimately, the goal is to ensure that a user who does not actively seed the RNG is highly likely to unique sequences in different interpreter processes, regardless of what time they started or what computer they are running on.  (It is important to note that different seeds do not guarantee non-overlapping sequences for all RNGs, though Mersenne Twister seems to be pretty good in this regard.)

Sufficient entropy from /dev/urandom should be enough on its own (though available might vary across platforms).  Lacking that timestamp XORed with some value unlikely to be shared by two processes is sufficient.